docs/tech-specs/basic-spec.md

```diff
@@ -6,6 +6,280 @@
 | *By* | Imran M Yousuf |
 | *Date* | November 10, 2020 |
 
-## Problem Statement / Goals
+## Problem Statement / Goal
 
-In a service oriented or microservice architecture the necessity to reliable pass messages between systems is essential. We want a reliable message broker without increasing architectural complexity.
+In a service oriented or microservice architecture the necessity to reliably pass messages between systems is essential. We want a reliable message broker without increasing architectural complexity with high throughput and scalability.
+
+## Assumptions & Considerations
+
+The assumptions into this tech spec are -
+
+* All the consumers have an HTTP based interface available
+* The interface or data model of webhook payload is agreed upon between producer and consumer
+* The consumer will take care of out of order message being pushed
+
+The considerations are -
+
+* The broker will ensure at least once webhook event is delivered
+* Consumer should have access to all the messages it failed to receive due to error on its part
+* Consumer could access past messages for replaying purpose
+
+Key attributes deliberately out of consideration -
+
+* There will be no guarantee of message order
+* There is no consideration for Consumer errors caused by bugs
+
+## Use cases
+
+The obvious goal is to serve the purpose of Enterprise Service Bus over HTTP; it is best suited when ordering is not a necessity; for example, change streams with optimistic locking. This Broker is not intended to be used as a replacement for Kafka or AWS Kinesis when order sequence is absolutely necessary or is of streaming nature with funnel effect; in addition to streaming it is also not intended to be used for batch processing purpose.
+
+### Yet another message broker
+
+The key complexity that brokers such as SQS, AMQP based broker, Celery, Kinesis, Kafka is it needs a separate worker process that needs to be deployed in addition to your Web API. On the surface it might seem not so significant; but when we think about infrastructure code for deploying Web API **and** the worker; plus it also means we need to add 2 different kind of monitoring, logging, scalability and fail-over code to solve; and depending on volume the separate infrastructure may not make sense either. With auto-scalable infrastructure, I would argue that, most projects can do with the workers being HTTP to begin with.
```

Comment:
`Celery` with eager mode enabled doesn't need a separate worker process, afaik.

Reply:
Yes [Eager Mode](https://docs.celeryproject.org/en/stable/userguide/configuration.html#std-setting-task_always_eager) does not require a worker; but not sure why anyone would use eager mode in observer pattern at all!

Reply:
In addition Celery is `1:1`; where broker enables `1:*` more akin to pub/sub model.